src/pudl/helpers.py

```diff
@@ -190,52 +190,133 @@ def download_zip_url(
+    return area_name_col.str.lower().replace(to_replace=replace_dict, regex=True)
+
 
-    # force the code columns to be nullable strings - the leading zeros are
-    # important
-    df = df.astype({"state_id_fips": pd.StringDtype()})
+def add_state_id_fips(
+    df: pd.DataFrame, geocodes: pd.DataFrame, state_col: str
+) -> pd.DataFrame:
+    """Add the State FIPS codes.
+
+    First we need to merge in the state abbreviation to the fips codes
 
+    Note: There are state fips codes in this pudl-compiled POLITICAL_SUBDIVISIONS
+    going with using the census based codes instead of our compiled codes even though
+    I checked they are all the same. Still feels better to sue a more og source.
+    """
+    state_abbr = (
+        POLITICAL_SUBDIVISIONS.loc[:, ["subdivision_code", "subdivision_name"]]
+        .rename(
+            columns={"subdivision_code": state_col, "subdivision_name": "area_name"}
+        )
+        .assign(area_name_tmp=lambda x: _clean_area_name_col(x["area_name"], {}))
+    )
+    states = (
+        geocodes.loc[geocodes["fips_level"] == "040", ["area_name", "state_id_fips"]]
+        .drop_duplicates()
+        .assign(area_name_tmp=lambda x: _clean_area_name_col(x["area_name"], {}))
+        .merge(state_abbr, on=["area_name_tmp"], how="left", validate="1:1")[
+            [state_col, "state_id_fips"]
+        ]
+    )
+    df = df.merge(states, on=state_col, how="left", validate="m:1")
     logger.info(
         f"Assigned state FIPS codes for "
-        f"{len(df[df.state_id_fips.notnull()]) / len(df):.2%} of records."
+        f"{len(df[df.state_id_fips.notnull()]) / len(df):.5%} of records."
     )
-    if county_col:
-        df["county_id_fips"] = df.apply(
-            lambda x: (
-                af.get_county_fips(state=x[state_col], county=x[county_col])
-                if pd.notnull(x[county_col]) and pd.notnull(x[state_col])
-                else pd.NA
-            ),
-            axis=1,
-        )
-        # force the code columns to be nullable strings - the leading zeros are
-        # important
-        df = df.astype({"county_id_fips": pd.StringDtype()})
-        logger.info(
-            f"Assigned county FIPS codes for "
-            f"{len(df[df.county_id_fips.notnull()]) / len(df):.2%} of records."
+    return df
+
+
+def add_county_fips_id(
+    df: pd.DataFrame, geocodes: pd.DataFrame, county_col: str
+) -> pd.DataFrame:
+    """Add the County FIPS codes to a table with State FIPS codes."""
+    diacretics = {
+        r"ñ": "n",
+        r"'": "",
+        r"ó": "o",
+        r"í": "i",
+        r"á": "a",
+        r"ü": "u",
+        r"é": "e",
+        r"î": "i",
+        r"è": "e",
+        r"à": "a",
+        r"ì": "i",
+        r"å": "a",
+    }
+    abbrevs = {"ft. ": "fort ", "st. ": "saint ", "ste. ": "sainte "}
+
+    county_types = {
+        r" (county|city|city and borough|borough|census area|municipio|municipality|district|parish)$": ""
+    }
```

Comment:
i liberally copy + pasted this from [here](https://github.com/fitnr/addfips/blob/c7f96a2b4906c97219d0003080b9779b2ae464cd/src/addfips/addfips.py#L28)